### PR TITLE
Change aot_module_simplified to take take arguments directly

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89773
* #89672
* #89671
* #89670
* __->__ #89669
* #89736

This is extracted from voz's #89392

Previously, the implementation did some half-assed caching where it
returned a callable, that when invoked for the first time, actually
performed the compilation.  Delaying the compilation like this...
seems totally unnecessary?  To make matters worse, this has cost
(we have to check if we hit the cache) and unsound (because the
compiled function may not be valid for other arguments.)

So instead, we ask user to provide arguments, and compile everything
immediately.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
failure here looks legit

**Review by voznesenskym — APPROVED** [review]

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -1735,6 +1735,7 @@ def forward(self, *args, **kwargs):
 
 def aot_module_simplified(
     mod: nn.Module,
+    args,
```
Comment (voznesenskym) [inline functorch/_src/aot_autograd.py:1738]: give it a type please, even if its Optional[List[Any]]

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -1795,27 +1796,30 @@ def functional_call(*args, **kwargs):
...
-                aot_config,
-            )
-        return compiled_fn(args)
-
-    def forward(*args):
-        return compiled_f(
-            *params_flat,
-            *args,
-        )
+    compiled_fn = create_aot_dispatcher_function(
```
Comment (voznesenskym) [inline functorch/_src/aot_autograd.py:1803]: can you underscore `create_aot_dispatcher_function` like in the source PR?

**Review by voznesenskym — APPROVED** [review]

File: functorch/_src/aot_autograd.py (hunk 2 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1803]: You're welcome to underscore functions in a follow up PR, there's a lot more identifiers than `create_aot_dispatcher_function` that need underscoring, and underscoring the function here contributed unnecessary line noise to the PR

Comment (ezyang) [reply functorch/_src/aot_autograd.py:1803]: In general, the functorch convention is that all identifiers in `functorch/_src` are considered private (even if the identifier doesn't have a leading underscore), and public identifiers are selectively reexported in `functorch` itself). So it is unnecessary to mark identifiers private on a one-by-one basis.

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1738]: I would like to give it the type `List[Tensor]` but as there may be SymInt inputs I am not sure it is accurate.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -1934,27 +1935,30 @@ def functional_call(*args, **kwargs):
...
+        functional_call,
+        full_args,
+        aot_config,
+    )
 
+    # TODO: There is something deeply wrong here; compiled_fn running with
```
Comment (Chillee) [inline functorch/_src/aot_autograd.py:1948]: I don't think `compiled_fn` actually returns the boxed calling convention (`compiled_fn` here should be the `autograd.Function` object, not the callable returned from the backend.

**Review by Chillee — APPROVED** [review]
Discussed this change with Voz previously, makes sense.

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1948]: It gets boxed up here 

```
        if needs_autograd:
            return make_boxed_func(
                aot_dispatch_autograd(flat_fn, fake_flat_tensor_args, aot_config)
            )
```

**Pull request closed without merging**